pkg/store/proxy.go

```diff
@@ -40,24 +44,51 @@ type ProxyStore struct {
 	stores         func() []Client
 	component      component.StoreAPI
 	selectorLabels labels.Labels
+
+	responseTimeout        time.Duration
+	streamResponseDuration *prometheus.HistogramVec
+	streamDuration         *prometheus.HistogramVec
 }
 
 // NewProxyStore returns a new ProxyStore that uses the given clients that implements storeAPI to fan-in all series to the client.
 // Note that there is no deduplication support. Deduplication should be done on the highest level (just before PromQL)
 func NewProxyStore(
 	logger log.Logger,
+	reg prometheus.Registerer,
 	stores func() []Client,
 	component component.StoreAPI,
 	selectorLabels labels.Labels,
+	responseTimeout time.Duration,
 ) *ProxyStore {
+	streamResponseDuration := prometheus.NewHistogramVec(prometheus.HistogramOpts{
```

Comment:
I am sorry, I don't really understand what exactly this measures in terms of Querier usage. It is all gRPC streams? Is it just StoreAPI series? It is not duplicated with what gRPC gives natively via grpc middleware (interceptor) we define in `cmd/thanos/...` code?

And what is the difference between two? Single read from stream? You mean single message/stream packet?

And  the other thing consume gRPC stream? What's the difference between consume vs read? 

I think we need to:
* check if that is not duplicated in gRPC client interceptor
* name it better and document better

Reply:
> And what is the difference between two? Single read from stream? You mean single message/stream packet?

Yeah single response from grpc's `Recv()` , which was the issue in my outage.

>And what is the difference between two? Single read from stream? You mean single message/stream packet?

One is for the whole stream to finish, another for a single `Recv()` operation to finish.

> check if that is not duplicated in gRPC client interceptor

It's not, but the whole stream might be duplicated. Need check if we can include store address labels there.

https://github.com/grpc-ecosystem/go-grpc-prometheus/blob/master/client_metrics.go#L159-L170
https://github.com/grpc-ecosystem/go-grpc-prometheus/blob/c225b8c3b01faf2899099b768856a9e916e5087b/client_reporter.go#L34

> name it better and document better

Suggestions?

Reply:
Yea I would merge this PR without the metrics (to not block it), and improve metrics to grpc middleware package at some point. (: Thoughts?